runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java

```diff
@@ -1110,17 +1109,7 @@ public void updateWatermarks(
     pendingUpdates.offer(
         PendingWatermarkUpdate.create(
             executable, completed, timerUpdate, unprocessedInputs, outputs, earliestHold));
-    tryApplyPendingUpdates();
-  }
-
-  private void tryApplyPendingUpdates() {
-    if (refreshLock.tryLock()) {
-      try {
-        applyNUpdates(MAX_INCREMENTAL_UPDATES);
-      } finally {
-        refreshLock.unlock();
-      }
-    }
```

Comment:
Not sure what was the purpose for limiting the number of applied updates, but that seems to be wrong - we need to apply all updates at all times, because otherwise we again might not hold the downstream watermark back enough.

Reply:
Updates can only move watermarks forward. I think this is just a rate limiting effect. This is from #1287. I believe it is correct. Whenever a bundle is completed, it can allow the watermark to move forward - but not past other holds. In this way, adding updates does not need a shared lock with applying and removing updates.

Reply:
Agree, I misread the code, there is force apply of all updates triggered from `EvaluationContext#extractFiredTimers()`, which is correct. My tests still pass even after returning this code back.